draft-ietf-dprive-dnsoquic.md

```diff
@@ -412,8 +412,9 @@ the client will establish a new connection and send the query over that
 connection.
 
 Clients MAY discard their connection to the server before the idle timeout
-expires. If they do that, they SHOULD close the connection explicitly, using
-QUIC's CONNECTION_CLOSE mechanism, and use the DoQ error code DOQ_NO_ERROR.
+expires. If they do that while responses to already sent queries are still expected,
```

Comment:
How about,

> A client that has outstanding queries SHOULD close the connection expliciting, using QUIC's CONNECTION_CLOSE mechanism and the DoQ error code DOQ_NO_ERROR.

Reply:
Closing the connection implicitly close all existing streams, so there is really no need to do that for each stream individually.

Reply:
The whole RTO thing is completely unnecessary.  Servers have to deal with not being able to retransmit packets to clients that disappear because clients disappear all the time.  Retransmissions of responses should not result in any more work for the server than sending the packets; it definitely shouldn't involve replaying any actions involved.

Requiring that clients individually cancel outstanding queries is unlikely to encourage compliance.  Even asking clients to send CONNECTION_CLOSE is not going to assure client behaviour.  Clients will drop their socket and disappear if that is in their interest.

As for load on servers, the effect of a client walking away should be no different than a connection hitting its idle timeout.  The server carries connection state until that time and then can reap it as it sees fit.  Connection state isn't free, but it's a pretty small amount of RAM that won't be accessed until cleanup is needed.

Reply:
My concern is some client implementation walking away from the connection immediately after receiving the response. I am fully aware that servers can deal will that; at the same time, the code path to deal with that is something like "repeat the last packets three times on timer", which is not so good. Should we say something?

I get that we are already telling the client to keep connections alive for an idle timer, so the connection is there if a new transaction is required. But we are looking at "alternative" implementations that don't do that. DNS UDP implementations deal with that just fine, because the server never repeats anything. It pains me somewhat that QUIC implementations will be worse...

Reply:
This is something HTTP implementations have dealt with for ages.  No doubt DNS TCP implementations have to manage the same problem too.  TCP RST is easier to produce, but QUIC has a reset equivalent that could work.

I'd say that you should just get comfortable with the discomfort.  Unless you think that there is a feature that you might negotiate in QUIC that alleviates it somehow.  I can't think of anything.

For me, I see this as a really tiny cost when compared to the other benefits of the protocol.  I don't think that it needs to be addressed.

Reply:
I don't think that telling such clients to not walk away is realistic, but I do think that such clients could be convinced to just close the connection instead of walking away. I also think that if a client closes the connection just after receiving the data it wanted, that just one packet sent by the client (ack and close), instead of three or four repeat attempts by the server. So I believe that's the right recommendation.